### PR TITLE
Document aws json rpc protocols

### PR DESCRIPTION
**Pull request by JordonPhillips**
This fleshes out the documentation for the `awsJson1_0` protocol.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 1]
```diff
@@ -61,4 +61,208 @@ See
...
+        For example, the value for the operation ``ns.example#MyOp`` of the
+        service ``ns.example#MyService`` is ``MyService.MyOp``.
+
+
+-------------
+HTTP Bindings
+-------------
+
+Every request for the ``awsJson1_0`` protocol MUST be sent to the
+root URL (``/``) using the HTTP POST method.
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:120]: Would have expected \`\`Post\`\` here.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 2]
```diff
@@ -61,4 +61,208 @@ See
...
+The ``awsJson1_0`` protocol supports the following traits
+that affect serialization:
+
+.. list-table::
+    :header-rows: 1
+    :widths: 20 80
+
+    * - Trait
+      - Description
+    * - :ref:`jsonName <jsonName-trait>`
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:78]: Does this need to have `@cors` listed?

I think this supports `@timestampFormat`, even if services don't use it yet.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 3]
```diff
@@ -61,4 +61,208 @@ See
...
+          - By default, the JSON property names used in serialized structures are
+            the same as a structure member name. The ``jsonName`` trait changes
+            the JSON property name to a custom value.
+
+
+----------------
+Protocol Headers
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:85]: What about merging this section and `HTTP Bindings` in to one `Protocol Behavior` section? I think leading with the content down in `HTTP Bindings` is the most important, then followed by these header notes.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 4]
```diff
@@ -61,4 +61,208 @@ See
...
+
+-------------
+HTTP Bindings
+-------------
+
+Every request for the ``awsJson1_0`` protocol MUST be sent to the
+root URL (``/``) using the HTTP POST method.
+
+The ``awsJson1_0`` protocol does not support custom HTTP bindings.
+HTTP binding traits MUST be ignored if they are present.
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:123]: "HTTP binding traits" should probably link to that page of the spec so implementers can know.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 5]
```diff
@@ -61,4 +61,208 @@ See
...
+      - JSON object. A union is serialized identically as a ``structure``
+        shape, but only a single member can be set to a non-null value.
+
+
+~~~~~~~~~~~~~~~~~~~~~~~~
+Empty body serialization
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+When an operation has no modeled input or if no input parameters are provided,
+a client MUST send an empty JSON object (``{}``) as input. Server-side
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:197]: This line's "input" should probably be "the request body".

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 6]
```diff
@@ -61,4 +61,208 @@ See
...
+implementations MUST also accept empty request bodies.
+
+When an operation has no modeled output or no output parameters are provided,
+server implementations SHOULD send an empty body. The server MUST still send
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:201]: "empty body" -> "empty response body"

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 7]
```diff
@@ -61,4 +61,208 @@ See
...
+the protocol's ``Content-Type`` header in this case. Clients MUST also accept
+an empty JSON object.
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:203]: "empty JSON object" -> "empty JSON object as the response body"

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 8]
```diff
@@ -61,4 +61,208 @@ See
...
+
+-----------------------------
+Operation error serialization
+-----------------------------
+
+Error responses in the ``awsJson1_0`` protocol are serialized identically to
+standard responses with one additional component to distinguish which error
+is contained. The component MUST be one of the following: an additional header
+with the name ``X-Amzn-Errortype``, a body field with the name ``code``, or a
+body field named ``__type``. New server-side protocol implementations SHOULD
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:216]: Can this SHOULD be moved in front of the MUST? Server implementations should see that information first.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 9]
```diff
@@ -61,4 +61,208 @@ See
...
+Differences between ``awsJson1_0`` and ``awsJson1_1``
+-----------------------------------------------------
+
+There are two difference between ``awsJson1_0`` and
+:ref:`aws.protocols#awsJson1_1-trait`. First, the static value of the
+``Content-Type`` header is different. In ``awsJson1_0`` it MUST be
+``application/x-amz-json-1.0`` and in ``awsJson1_1`` it MUST be
+``application/x-amz-json-1.1``.
+
+Second, error type serialization is different. In ``awsJson1_0`` servers SHOULD
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:248]: "In \`\`awsJSON1_0\`\`, servers", need a comma inserted.

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 10]
```diff
@@ -61,4 +61,208 @@ See
...
+send the error shape's full :ref:`shape-id` in the ``__type`` field of the body.
+In ``awsJson1_1`` servers SHOULD only send the error's
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:250]: "In \`\`awsJSON1_1\`\`, servers", need a comma inserted.

**Review by kstich — CHANGES_REQUESTED** [review]

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 11]
```diff
@@ -61,4 +61,208 @@ See
     This protocol only permits the :ref:`httpPayload-trait` to be applied to
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:61]: This protocol doesn't support `@httpPayload`

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst (hunk 1 above)
Comment (mtdowling) [reply docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:120]: Do we use backticks everywhere else for HTTP methods? If not, I think omitting backticks is fine.

Comment (kstich) [reply docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:120]: We use quotes elsewhere it looks like:

>The \`\`method\`\` property defines the HTTP method of the operation (e.g., "GET",
"PUT", "POST", "DELETE", "PATCH", etc).

**Comment by JordonPhillips** [comment]
Updated to also include awsJson1_1 using a template with substitutions to avoid duplicating text

File: docs/source/1.0/spec/aws/aws-json-1_1-protocol.rst [hunk 12]
```diff
@@ -56,9 +56,8 @@ See
...
         }
 
-.. important::
-
-    This protocol only permits the :ref:`httpPayload-trait` to be applied to
-    members that target structures, documents, strings, blobs, or unions.
-
-*TODO: Add specifications, protocol examples, etc.*
+.. |quoted shape name| replace:: ``awsJson1_1``
+.. |protocol content type| replace:: ``application/x-amz-json-1.0``
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_1-protocol.rst:60]: Should be `1.1`

File: docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst [hunk 13]
```diff
@@ -19,7 +19,7 @@ This specification defines the ``aws.protocols#awsJson1_0`` protocol.
 ----------------------------------
 
 Summary
-    Adds support for an HTTP protocol that sends POST requests and
+    Adds support for an HTTP protocol that sends "POST" requests and
```
Comment (kstich) [inline docs/source/1.0/spec/aws/aws-json-1_0-protocol.rst:22]: Should do this in the `awsJson1_1` docs?

**Review by kstich — CHANGES_REQUESTED** [review]

**Review by kstich — APPROVED** [review]

**Pull request merged**